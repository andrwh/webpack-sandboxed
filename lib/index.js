// @flow

export { WebpackRunner as default } from './webpack'
