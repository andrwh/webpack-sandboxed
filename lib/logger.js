// @flow

const prefix = 'webpack-sandboxed'
let debug

try {
  debug = require('debug')
} catch(e) {
  debug = () => () => undefined
}

type Logger = {
  debug: (...message: Array<mixed>) => void;
  info: (...message: Array<mixed>) => void;
  warn: (...message: Array<mixed>) => void;
  error: (...message: Array<mixed>) => void;
}

export default function (name: string): Logger {
  return {
    debug: debug(`${prefix}:debug:${name}`),
    info: debug(`${prefix}:info:${name}`),
    warn: debug(`${prefix}:warn:${name}`),
    error: debug(`${prefix}:error:${name}`)
  }
}
