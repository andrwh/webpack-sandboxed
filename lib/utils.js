// @flow

import fs from 'fs'
import path from 'path'

export function fsReaddir(dirpath: string) {
  return new Promise((resolve, reject) => {
    fs.readdir(dirpath, (error, files) => {
      if (error) return reject(error)
      else return resolve(files)
    })
  })
}

export function fsReadFile(filepath: string) {
  return new Promise((resolve, reject) => {
    fs.readFile(filepath, (error, data) => {
      if (error) return reject(error)
      else return resolve(data)
    })
  })
}

export function fsStat(dirpath: string): Promise<fs.Stats> {
  return new Promise((resolve, reject) => {
    fs.stat(dirpath, (error, stats) => {
      if (error) return reject(error)
      else return resolve(stats)
    })
  })
}

export function fsPipe(inStream: fs.ReadStream, outStream: fs.WriteStream): Promise<void> {
  return new Promise((resolve, reject) => {
    inStream.pipe(outStream)
    outStream.on('finish', () => resolve())
    outStream.on('error', error => reject(error))
  })
}

export async function walkDirectory(root: string, callback: ?(file: string) => Promise<void>): Promise<Array<string>> {
  let files = []
  try {
    files = await fsReaddir(root)
    if (files.length === 0) return []
  } catch (e) {
    if (e.code === 'ENOENT') return []
    else throw e
  }

  let filePaths = []

  for (let file of files) {
    const filePath = path.resolve(root, file)
    const stats = await fsStat(filePath)

    if (stats.isDirectory()) {
      await walkDirectory(filePath, callback)
    } else {
      filePaths = filePaths.concat(filePath)
      if (callback) await callback(filePath)
    }
  }
  return filePaths
}

const IGNORE_DEFAULT = [/\.git/, /dist/]

export async function findDirectory(root: string, name: string, ignore: Array<RegExp> = IGNORE_DEFAULT): Promise<Array<string>> {
  if (root.endsWith(path.basename(name))) return [root]

  let files = []
  try {
    files = await fsReaddir(root)
    if (files.length === 0) return []
  } catch (e) {
    if (e.code === 'ENOENT') return []
    else throw e
  }

  const fileInfos = await Promise.all(files
    .map(file => path.resolve(root, file))
    .map(async (file) => {
      const stats = await fsStat(file)
      const isDirectory = stats.isDirectory()
      return {file, isDirectory}
    }))

  const isNotIgnored = (file: string) => ignore.reduce((ignore, pattern) => ignore && !pattern.test(file), true)

  const subDirs = fileInfos
    .filter(info => info.isDirectory && isNotIgnored(info.file))
    .map(info => info.file)

  const moreDirs = await Promise.all(subDirs.map(async (dir) => await findDirectory(dir, name)))

  return moreDirs.reduce((dirs, dir) => dirs.concat(dir), [])
}

export function stringify(value: mixed, escape: ?string): string {
  if (value === null || typeof value === 'undefined'){
    return ''
  } else if (value instanceof RegExp) {
    return value.toString()
  } else if (Array.isArray(value)) {
    return `[${value.map(e => stringify(e, escape)).join(',')}]`
  } else if (typeof value === 'object') {
    let string = '{'
    const keys = Object.keys(value)

    for (let i = 0; i < keys.length; i += 1) {
      const key = keys[i]
      string += `${key}: ${stringify(value[key], escape)}`
      if (i < keys.length - 1) string += ', '
    }
    return string + '}'
  } else if (typeof value === 'string' && typeof escape === 'string' && value.startsWith(escape)) {
    return value.substring(escape.length)
  } else {
    return JSON.stringify(value)
  }
}
