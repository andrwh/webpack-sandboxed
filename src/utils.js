// @flow

import fs from 'fs'
import path from 'path'

export function fsReaddir(dirpath: string) {
  return new Promise((resolve, reject) => {
    fs.readdir(dirpath, (error, files) => {
      if (error) return reject(error)
      else return resolve(files)
    })
  })
}

export function fsReadFile(filepath: string) {
  return new Promise((resolve, reject) => {
    fs.readFile(filepath, (error, data) => {
      if (error) return reject(error)
      else return resolve(data)
    })
  })
}

export function fsStat(dirpath: string): Promise<fs.Stats> {
  return new Promise((resolve, reject) => {
    fs.stat(dirpath, (error, stats) => {
      if (error) return reject(error)
      else return resolve(stats)
    })
  })
}

export async function walkDirectory(root: string, callback: ?(file: string) => Promise<void>): Promise<Array<string>> {
  let files = []
  try {
    files = await fsReaddir(root)
    if (files.length === 0) return []
  } catch (e) {
    if (e.code === 'ENOENT') return []
    else throw e
  }

  let filePaths = []

  for (let file of files) {
    const filePath = path.resolve(root, file)
    const stats = await fsStat(filePath)

    if (stats.isDirectory()) {
      await walkDirectory(filePath, callback)
    } else {
      filePaths = filePaths.concat(filePath)
      if (callback) await callback(filePath)
    }
  }
  return filePaths
}

const IGNORE_DEFAULT = [/\.git/, /dist/]

export async function findDirectory(root: string, name: string, ignore: Array<RegExp> = IGNORE_DEFAULT): Promise<Array<string>> {
  if (root.endsWith(path.basename(name))) return [root]

  let files = []
  try {
    files = await fsReaddir(root)
    if (files.length === 0) return []
  } catch (e) {
    if (e.code === 'ENOENT') return []
    else throw e
  }

  const fileInfos = await Promise.all(files
    .map(file => path.resolve(root, file))
    .map(async (file) => {
      const stats = await fsStat(file)
      const isDirectory = stats.isDirectory()
      return {file, isDirectory}
    }))

  const isNotIgnored = (file: string) => ignore.reduce((ignore, pattern) => ignore && !pattern.test(file), true)

  const subDirs = fileInfos
    .filter(info => info.isDirectory && isNotIgnored(info.file))
    .map(info => info.file)

  const moreDirs = await Promise.all(subDirs.map(async (dir) => await findDirectory(dir, name)))

  return moreDirs.reduce((dirs, dir) => dirs.concat(dir), [])
}
