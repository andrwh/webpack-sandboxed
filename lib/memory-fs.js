// @flow

import fs from 'fs';
import path from 'path';
import requireResolve from 'resolve';
import MemoryFS from 'memory-fs';
import logger from './logger';
import * as utils from './utils';

export type { MemoryFS } from 'memory-fs';
type Arguments = {
  packages?: Array<string>,
  includes?: Array<string>,
  root?: string
};

const log = logger('memory-fs');

class PackageLoader {
  root: string;
  memfs: MemoryFS;
  cache: { [key: string]: boolean };

  constructor(root: string, memfs: MemoryFS) {
    this.root = root;
    this.memfs = memfs;
    this.cache = {};
  }

  resolvePackage(packageName: string): Promise<string> {
    return new Promise((resolve, reject) => {
      requireResolve(
        packageName,
        {
          // Use the path of package.json as fallback for the main file.
          packageFilter: (pkg, pkgJson) =>
            pkg.main ? pkg : Object.assign({}, pkg, { main: pkgJson })
        },
        (error, result) => error ? reject(error) : resolve(result)
      );
    })
      // Extract the package path (the first path segment after node_modules).
      .then(mainFilePath => /(^.*node_modules\/[^\/]+)/.exec(mainFilePath)[1]);
  }

  async loadPackage(packageName: string): Promise<void> {
    // Don't resolve the same dependency twice.
    if (this.cache[packageName]) return;
    else this.cache[packageName] = true;

    const packagePath = await this.resolvePackage(packageName);
    log.debug('Loading %s', packagePath);

    await utils.walkDirectory(packagePath, async file => {
      const match = /^.*(node_modules.+)$/.exec(file)[1];
      const exportPath = path.join(this.root, match);

      this.memfs.mkdirpSync(path.dirname(exportPath));
      await utils.fsPipe(
        fs.createReadStream(file),
        this.memfs.createWriteStream(exportPath)
      );

      // Also recursively load all dependencies of the loaded module.
      // TODO: optimize
      if (file === path.join(packagePath, 'package.json')) {
        const fileContent = await utils.fsReadFile(file);
        const { dependencies } = JSON.parse(fileContent.toString());
        if (dependencies) {
          await Promise.all(
            Object.keys(dependencies)
              .map(async dependency => await this.loadPackage(dependency))
          );
        }
      }
    });
  }

  async loadFile(inputPath: string, outputPath: string) {
    const stats = await utils.fsStat(inputPath);

    // If input is a file, load only this file.
    if (stats.isFile()) {
      const resolvedOutputPath = path.resolve(this.root, outputPath);
      this.memfs.mkdirpSync(path.dirname(resolvedOutputPath));
      await utils.fsPipe(
        fs.createReadStream(inputPath),
        this.memfs.createWriteStream(resolvedOutputPath)
      );
      return;
    }

    // Otherwise recursively load the directory.
    await utils.walkDirectory(inputPath, async file => {
      const relativeInputPath = path.relative(inputPath, file);
      const resolvedOutputPath = path.resolve(
        this.root,
        outputPath,
        relativeInputPath
      );

      this.memfs.mkdirpSync(path.dirname(resolvedOutputPath));
      await utils.fsPipe(
        fs.createReadStream(file),
        this.memfs.createWriteStream(resolvedOutputPath)
      );
    });
  }
}

export async function createInstance(args: Arguments = {}): Promise<MemoryFS> {
  const root = args.root || process.cwd();
  const memfs = new MemoryFS();
  const loader = new PackageLoader(root, memfs);
  const packages = args.packages || [];

  log.debug('Loading packages...');
  await Promise.all(packages.map(async p => await loader.loadPackage(p)));

  const includes = args.includes || [];

  log.debug('Loading includes...');
  await Promise.all(
    includes.map(
      async include =>
        await loader.loadFile(include, path.join(root, path.basename(include)))
    )
  );

  return memfs;
}
