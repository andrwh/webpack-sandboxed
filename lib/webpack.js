// @flow

import vm from 'vm';
import path from 'path';
import webpack from 'webpack';
import logger from './logger';
import * as memoryfs from './memory-fs';
import * as utils from './utils';

type WebpackBundle = { [key: string]: string };
type WebpackStats = {
  hasErrors: () => boolean,
  hasWarnings: () => boolean,
  toJson: (options: ?Object) => Object,
  toString: (options: ?Object) => Object
};
type WebpackTarget =
  | 'web'
  | 'webworker'
  | 'node'
  | 'async-node'
  | 'node-webkit'
  | 'electron';
type WebpackConfig = {
  context?: string,
  entry?: string | { [key: string]: string },
  context?: string,
  output?: {
    path: string,
    filename: string
  },
  module?: {
    loaders: Array<Object>
  },
  target?: WebpackTarget,
  plugins?: Array<Object> | '@plugins'
};
type OnComplete = (
  error: ?Error,
  bundle: ?WebpackBundle,
  stats: ?WebpackStats
) => void;
type Options = {
  config?: WebpackConfig,
  packages?: Array<string>,
  includes?: Array<string>
};

const log = logger('webpack');

export class SandboxVM {
  globals: Object;

  constructor(globals: Object = {}) {
    this.globals = globals;
  }

  run(source: string, onComplete: OnComplete): any {
    try {
      const script = new vm.Script(source, { displayErrors: true });
      const sandbox = Object.assign({}, this.globals, { onComplete });
      const context = vm.createContext(sandbox);
      return script.runInContext(context);
    } catch (error) {
      return onComplete(error);
    }
  }
}

export class WebpackRunner {
  vm: SandboxVM;
  memfs: memoryfs.MemoryFS;
  entry: string;
  script: string;

  constructor(vm: SandboxVM, memfs: memoryfs.MemoryFS, entry: string) {
    this.vm = vm;
    this.memfs = memfs;
    this.entry = entry;

    // The webpack script to run in the sandbox VM. Relies on global variables to be present.
    this.script = `
    const outDir = webpackConfig.output.path
    const compiler = webpack(webpackConfig)
    compiler.inputFileSystem = memfs
    compiler.resolvers.normal.fileSystem = memfs
    compiler.resolvers.context.fileSystem = memfs
    compiler.outputFileSystem = memfs
    compiler.run((error, stats) => {
      if (error) return onComplete(error)
      const files = memfs.readdirSync(outDir)
      const bundle = files.reduce((bundle, file) =>
        Object.assign(bundle, {[file]: memfs.readFileSync(path.join(outDir, file)).toString()})
      , {})
      memfs.rmdirSync(outDir)
      onComplete(null, bundle, stats)
    })
    `;
  }

  static async createInstance(options: Options = {}): Promise<WebpackRunner> {
    const packages = options.packages || [];
    const includes = options.includes || [];
    const config = options.config || {};

    // Create webpack configuration.
    // The root directory must match the real installation directory so that packages
    // like Babel can resolve other modules. The node module system in the sandbox
    // operates on the real file system and not on memory-fs.
    // We use the base path of the memory-fs module that should be installed
    // alongside the webpack-sandboxed module and its dependants.
    const root = /^(.*)node_modules/.exec(require.resolve('memory-fs'))[1];
    const entry = path.join(root, 'entry.js');
    const output = path.join(root, 'bundle/[name]-[hash].min.js');
    const baseConfig: WebpackConfig = {
      context: path.dirname(entry),
      entry: `./${path.basename(entry)}`,
      output: {
        path: path.dirname(output),
        filename: path.basename(output)
      }
    };
    const webpackConfig = utils.deepAssign(baseConfig, config);

    const memfs = await memoryfs.createInstance({ packages, includes, root });
    memfs.mkdirpSync(path.dirname(output));

    // Initialize VM with global variables to be used in the webpack script.
    const vm = new SandboxVM({
      path,
      webpack,
      memfs,
      webpackConfig
    });

    return new WebpackRunner(vm, memfs, entry);
  }

  async __run(source: string, onComplete: OnComplete) {
    this.memfs.mkdirpSync(path.dirname(this.entry));
    this.memfs.writeFileSync(this.entry, source);
    this.vm.run(this.script, (error, bundle, stats) =>
      onComplete(error, bundle, stats));
  }

  async run(source: string): Promise<[WebpackBundle, WebpackStats]> {
    return new Promise(async (resolve, reject) => {
      log.debug('Executing script...');
      await this.__run(source, (error, bundle, stats) => {
        if (error) {
          log.error('Failed to execute script.');
          reject(error);
        } else if (stats && stats.hasErrors()) {
          log.error('Script finished with errors.');
          reject(stats.toString());
        } else if (bundle && stats) {
          log.debug('Successfully compiled bundle.');
          resolve([bundle, stats]);
        } else {
          log.error('Failed to execute script.');
          reject(new Error('Unknown error.'));
        }
      });
    });
  }
}
