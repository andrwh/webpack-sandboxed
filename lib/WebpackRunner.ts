import path = require('path');
import webpack = require('webpack');
import logger from './logger';
import MemoryFS = require('memory-fs');

export type WebpackBundle = { [key: string]: string };
export type Options = {
  config?: webpack.Configuration;
  packages?: string[];
  includes?: string[];
  basedir?: string;
};

export interface WebpackBaseConfiguration extends webpack.Configuration {
  context: string;
  entry: string;
  output: {
    path: string;
    filename: string;
  };
  resolve: webpack.Resolve;
  resolveLoader: webpack.ResolveLoader;
}

type OnComplete = (error?: Error, bundle?: WebpackBundle, stats?: webpack.Stats) => void;

interface WebpackCompiler extends webpack.Compiler {
  inputFileSystem: any;
  resolvers: {
    normal: any;
    loader: any;
    context: any;
  };
}

const log = logger('webpack');

export default class WebpackRunner {
  private readonly memfs: MemoryFS;
  private readonly config: WebpackBaseConfiguration;

  constructor(memfs: MemoryFS, config: WebpackBaseConfiguration) {
    this.memfs = memfs;
    this.config = config;
  }

  private runAsync(source: string, onComplete: OnComplete): void {
    const entry = path.join(this.config.context, this.config.entry);
    this.memfs.mkdirpSync(path.dirname(entry));
    this.memfs.writeFileSync(entry, source);

    const outDir = this.config.output.path;
    const compiler = webpack(this.config) as WebpackCompiler;

    compiler.inputFileSystem = this.memfs;
    compiler.outputFileSystem = this.memfs;
    compiler.resolvers.normal.fileSystem = this.memfs;
    compiler.resolvers.loader.fileSystem = this.memfs;
    compiler.resolvers.context.fileSystem = this.memfs;

    compiler.run((error, stats) => {
      if (error) return onComplete(error);
      const files = this.memfs.readdirSync(outDir);
      const bundle = files.reduce(
        (object, file) =>
          Object.assign(object, {
            [file]: this.memfs.readFileSync(path.join(outDir, file)).toString()
          }),
        {}
      );
      this.memfs.rmdirSync(outDir);
      onComplete(undefined, bundle, stats);
    });
  }

  public run(source: string): Promise<[WebpackBundle, webpack.Stats]> {
    return new Promise((resolve, reject) => {
      log.debug('Executing script...');
      this.runAsync(source, (error, bundle, stats) => {
        if (error) {
          log.error('Failed to execute script.');
          reject(error);
        } else if (stats && stats.hasErrors()) {
          log.error('Script finished with errors.');
          reject(stats.toString());
        } else if (bundle && stats) {
          log.debug('Successfully compiled bundle.');
          resolve([bundle, stats]);
        } else {
          log.error('Failed to execute script.');
          reject(new Error('Unknown error.'));
        }
      });
    });
  }

  private nodeModulesInContext(): string[] {
    return this.memfs.readdirSync(path.join(this.config.context, 'node_modules'));
  }

  public toJSON(): object {
    return {
      config: Object.freeze(Object.assign({}, this.config)),
      node_modules: this.nodeModulesInContext().sort()
    };
  }

  public toString(): string {
    return JSON.stringify(this.toJSON(), null, 2);
  }
}
