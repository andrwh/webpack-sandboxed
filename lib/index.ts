import path = require('path');
import createMemoryFS from './memory-fs';
import { deepAssign, findNodeModulesPath } from './utils';
import WebpackRunner, { Options, WebpackBaseConfiguration } from './WebpackRunner';

export default async function createInstance(options: Options = {}): Promise<WebpackRunner> {
  const packages = options.packages || [];
  const includes = options.includes || [];
  const config = options.config || {};
  const nodeModulesPath = findNodeModulesPath();

  // Create webpack configuration.
  // The root directory must match the real installation directory so that tools
  // like Babel can resolve other modules. The node module system actually
  // operates on the real file system and not on memory-fs.
  const root = options.basedir || path.dirname(nodeModulesPath);
  const entry = path.join(root, 'entry.js');
  const output = path.join(root, 'bundle/[name].[chunkhash].js');
  const includesPath = path.join(root, 'includes');

  const baseConfig: WebpackBaseConfiguration = {
    context: path.dirname(entry),
    entry: `./${path.basename(entry)}`,
    output: {
      path: path.dirname(output),
      filename: path.basename(output)
    },
    resolve: {
      modules: [path.resolve('node_modules'), path.join(root, 'node_modules'), includesPath]
    },
    resolveLoader: {
      modules: [path.resolve('node_modules'), path.join(root, 'node_modules')]
    }
  };
  const webpackConfig = deepAssign(config, baseConfig) as WebpackBaseConfiguration;

  // Webpack buildins like global, amd-define etc. are required.
  const webpackPath = require.resolve('webpack') && path.join(nodeModulesPath, 'webpack');
  const webpackBuildinPath = path.join(webpackPath, 'buildin');
  packages.push(webpackBuildinPath);
  // This webpack dependency includes node APIs required by other modules, e.g. loaders.
  if (require.resolve('node-libs-browser')) packages.push('node-libs-browser');

  const memfs = await createMemoryFS({ packages, includes, includesPath, root });
  memfs.mkdirpSync(path.dirname(output));

  return new WebpackRunner(memfs, webpackConfig);
}
