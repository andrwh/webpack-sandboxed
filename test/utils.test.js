// @flow

import * as utils from '../lib/utils'

describe('utils', () => {
  describe('stringify', () => {
    it('should stringify a complext object', async () => {
      const object = {
        hello: 'world',
        regex: /^foobar$/,
        obj: {
          arr: [1, 2, {test: true}]
        }
      }
      expect(utils.stringify(object)).toMatchSnapshot()
    })

    it('should escape special strings', async () => {
      const object = {
        foo: '@variable'
      }
      expect(utils.stringify(object, '@')).toBe('{foo: variable}')
    })
  })

  describe('deepAssign', () => {
    it('should deeply merge two objects', async () => {
      const objectA = {
        hello: 'test',
        a: 'a',
        obj: {
          x: 1,
          arr: [1, 2, [42]]
        }
      }
      const objectB = {
        hello: 'world',
        b: 'b',
        obj: {
          y: 2,
          arr: [3, 4, [42]]
        }
      }
      const merged = {
        hello: 'world',
        a: 'a',
        b: 'b',
        obj: {
          x: 1,
          y: 2,
          arr: [1, 2, [42], 3, 4, [42]]
        }
      }
      expect(utils.deepAssign({}, objectA, objectB)).toEqual(merged)
    })
  })
})
