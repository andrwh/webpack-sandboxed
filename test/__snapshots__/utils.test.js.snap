exports[`utils stringify should stringify a complext object 1`] = `"{hello: \"world\", regex: /^foobar$/, obj: {arr: [1,2,{test: true}]}}"`;
